### DISCUSSION TITLE
Metal and indirect light

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3495] [Q&A] [answered]
By playing with `material_sandbox`, I see that having a reasonable amount of indirect light is critical to making shiny metal look good; otherwise you see specular highlights and the rest is black.  But when I try adding ambient light without a cube map, it does not seem to have any effect on metal, though it works as expected on dielectrics.

```
    math::float3 ambientColor = { 1.0f, 1.0f, 1.0f };
    _indirLight = IndirectLight::Builder()
		.intensity( 1000000.0f )
		.irradiance( 1, &ambientColor )
		.build( *_engine );
    _scene->setIndirectLight( _indirLight );
```

Am I doing something wrong?

**Comment by romainguy** [comment, marked as answer]
irradiance is only used for diffuse lighting. Metals only have specular reflections, you need to provide a cubemap for metals to work. But that cubemap only needs to be a few pixels large and can be only a solid color.

**Reply by jwwalker** [reply]
That would be set using the `IndirectLight::Builder::reflections` method?  The docs for its cubemap parameter say "Each cubemap level encodes a the irradiance for a roughness level", so we have irradiance there too, which confuses me.

**Reply by romainguy** [reply]
Yes, by the way for irradiance() you must pass an array of 9*3 floats.

**Reply by jwwalker** [reply]
Wait, why do I need to pass 9*3 floats to `irradiance()`?  The docs say that the number of bands must be 1, 2, or 3, and the size of the array of `float3` is the square of that.  So you're saying that the number of bands can't be 1 and must be 3?  Under what conditions?

**Comment by jwwalker** [comment]
For benefit of any others reading this, here's working code that makes a simple ambient light that works on metal.

```
	static unsigned char oneWhitePixel[] = { 0xFF, 0xFF, 0xFF };
	backend::PixelBufferDescriptor whitePixBuf( oneWhitePixel, sizeof(oneWhitePixel),
		backend::PixelDataFormat::RGB, backend::PixelDataType::UBYTE );
	_whiteCube = Texture::Builder()
		.width(1)
		.height(1)
		.levels(1)
		.sampler( filament::backend::SamplerType::SAMPLER_CUBEMAP )
		.format( filament::backend::TextureFormat::RGB8 )
		.build( *_engine );
	backend::FaceOffsets offsets( 0 );
	_whiteCube->setImage( *_engine, 0, std::move(whitePixBuf), offsets );
	
	_indirLight = IndirectLight::Builder()
		.intensity( 20000.0f )
		.reflections( _whiteCube )
		.build( *_engine );

```
